src/uct/ib/rc/accel/rc_mlx5_common.c

```diff
@@ -218,6 +218,181 @@ void uct_rc_mlx5_iface_common_tag_cleanup(uct_rc_mlx5_iface_common_t *iface,
+    data->seg_count = ucs_align_up(config->dm.count, 8);
+    data->device    = uct_ib_iface_device(&iface->super);
+
+    dm_attr.length    = data->seg_len * data->seg_count;
+    dm_attr.comp_mask = 0;
+    data->dm          = ibv_exp_alloc_dm(data->device->ibv_context, &dm_attr);
+
+    if (data->dm == NULL) {
+        /* TODO: prompt warning? */
+        ucs_debug("ibv_exp_alloc_dm() error - On Device Memory allocation failed, %d %m", errno);
+        return UCS_OK;
+    }
+
+    memset(&mr_in, 0, sizeof(mr_in));
+    mr_in.pd           = uct_ib_iface_md(&iface->super)->pd;
+    mr_in.comp_mask    = IBV_EXP_REG_MR_DM;
+    mr_in.dm           = data->dm;
+    mr_in.length       = dm_attr.length;
+    data->mr           = ibv_exp_reg_mr(&mr_in);
+    if (data->mr == NULL) {
+        ucs_warn("ibv_exp_reg_mr() error - On Device Memory registration failed, %d %m", errno);
+        goto failed_mr;
+    } else {
+        data->start_va = ((uct_mlx5_dm_va_t*)data->dm)->start_va;
+    }
+
+    status = ucs_mpool_init(&data->mp, sizeof(uct_iface_dm_mp_priv_t),
+                            sizeof(uct_rc_iface_send_desc_t), 0, UCS_SYS_CACHE_LINE_SIZE,
+                            data->seg_count, data->seg_count,
+                            &uct_dm_iface_mpool_ops, "mlx5_dm_desc");
+    if (status != UCS_OK) {
+        goto failed_mpool;
+    }
+
+    uct_iface_dm_mp_priv(&data->mp)->dm    = data;
+    uct_iface_dm_mp_priv(&data->mp)->count = 0;
+
+complete:
+    /* DM initialization may fail due to any reason, just
+     * free resources & continue without DM */
+    return UCS_OK;
+
+failed_mpool:
+    ibv_dereg_mr(data->mr);
+failed_mr:
+    ibv_exp_free_dm(data->dm);
+    data->dm = NULL;
+    goto complete;
+}
+
+static void uct_mlx5_iface_cleanup_dm_tl(uct_mlx5_dm_data_t *data)
+{
+    if (data->dm) {
+        ucs_assert(data->mr != NULL);
+
+        ucs_mpool_cleanup(&data->mp, 1);
+        ibv_dereg_mr(data->mr);
+        ibv_exp_free_dm(data->dm);
+    }
+}
+#endif
+
+static ucs_status_t uct_mlx5_iface_init_dm(uct_rc_mlx5_iface_common_t *iface,
+                                           uct_rc_iface_t *rc_iface,
+                                           const uct_common_mlx5_iface_config_t *config)
+{
+#if HAVE_IBV_EXP_DM
+    if ((config->dm.seg_len * config->dm.count) == 0) {
+        iface->dm.dm = NULL;
+        return UCS_OK;
+    }
+
+    iface->dm.dm = uct_worker_tl_data_get(rc_iface->super.super.worker,
+                                          uct_rc_mlx5_dm_key,
+                                          uct_mlx5_dm_data_t,
+                                          uct_mlx5_iface_dm_device_cmp,
+                                          uct_mlx5_iface_init_dm_tl, rc_iface, config);
+    if (iface->dm.dm == NULL) {
+        return UCS_OK;
+    }
+
+    if (iface->dm.dm->dm == NULL) {
+        /* failed to initialize DM. put back object */
+        ucs_debug("put back allocated buffer");
```

Comment:
need to remote this debug? or add more info?

Reply:
removed